cpp/ql/test/query-tests/Likely Bugs/Memory Management/AllocaInLoop/AllocaInLoop2.c

```diff
@@ -0,0 +1,158 @@
+// semmle-extractor-options: --clang
+int printf(const char *restrict format, ...);
+int sprintf(char *restrict buf, const char *restrict format, ...);
+typedef unsigned long size_t;
+void *memcpy(void *restrict s1, const void *restrict s2, size_t n);
+void *malloc(size_t sz);
+void free(void *ptr);
+
+struct vtype { int i1, i2; };
+extern int w1, w2;
+
+void *_builtin_alloca(int sz);
+#define alloca __builtin_alloca
+
+// We forward-declare the Microsoft routines
+//_alloca and _malloca here.  Since they do not
+// originate from the <malloc.h> header, they
+// should not be flagged by our queries
+void *_alloca(size_t sz);
+void *_malloca(size_t sz);
+void _freea(void *ptr);
+
+#define NULL (void *)0
+
+// case 1: alloca called within a provably infinite loop
+void foo(const struct vtype* vec, int count) {
+	char iter;
+
+    do {
+        const struct vtype* v = vec + 2;
+        char *b1 = 0;
+        iter = 0;
+        if (b1 == NULL) {
+            if (w1 > w2) {
+                // Allocate the buffer on heap
+                b1 = (char *)malloc(w1);
+            } else {
+                // Allocate the buffer on stack
+                b1 = (char*) alloca(w1);  // [FLAG]
+                iter = 1;
+            }
+        }
+        memcpy(b1, v, w1);
+        printf("%s\n", b1);
+        if (w1 > w2) {
+        	free(b1);
+        }
+    } while (iter);
+}
+
+// case 2: alloca called within nested do-while(0) loops
+void bar(const struct vtype* vec, int count) {
+
+    do {
+        const struct vtype* v = vec + 2;
+        char *b1 = 0;
+        do {
+        if (b1 == NULL) {
+            if (w1 > w2) {
+                // Allocate the buffer on heap
+                b1 = (char *)malloc(w1);
+            } else {
+                // Allocate the buffer on stack
+                b1 = (char*) alloca(w1);  // [DO NOT FLAG]
```

Comment:
Please use our standard annotations instead of `// [FLAG]` and `// [DO NOT FLAG]`. They are: `// GOOD`, `// BAD`, `// GOOD [FALSE POSITIVE]`, or `// BAD [NOT DETECTED]`.